### PR TITLE
Sanitize dollar identifiers with leading zeros in `Identifier`.

### PR DESCRIPTION
**Pull request by MAJKFL**
This PR introduces sanitization of dollar identifiers with leading zeros to `Identifier`. 

The compiler doesn't treat dollar identifiers with leading zeros differently than the "canonical" ones, so for the following piece of code:

```swift
var foo = {
  print($0, $00000000, $1, $00001)
}

foo(4, 2)
```

The result will be:

```
4 4 2 2
```

This PR will also enable `SwiftLexicalLookup` to properly handle this edge case when introducing dollar identifiers as names in closures.

### CODE REVIEW

**Comment by MAJKFL** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/Identifier.swift [hunk 1]
```diff
@@ -26,41 +29,67 @@ public struct Identifier: Equatable, Hashable, Sendable {
...
-  public init(_ staticString: StaticString) {
-    self.raw = RawIdentifier(staticString)
-    self.arena = nil
-
-    let name = String(syntaxText: raw.name)
+  /// Create a new `Identifier` from given `staticString`.
+  ///
+  /// - Precondition: `staticString` is a canonical identifier i.e. doesn't
+  ///   use backticks and is not a dollar identifier with leading zeros.
+  public init(staticString: StaticString) {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/Identifier.swift:56]: What’s the reason why you added the `staticString` argument label?

**Review by ahoppen — COMMENTED** [review]
How did you find that, Jakub?

File: Sources/SwiftSyntax/Identifier.swift (hunk 1 above)
Comment (MAJKFL) [reply Sources/SwiftSyntax/Identifier.swift:56]: I noticed this used to produce a compilation error
```swift
let identifier = Identifier("smth") // Error: Ambiguous use of 'init(_:)'
```
Interestingly, these two worked absolutely fine
```swift
let x: StaticString = "smth"
let identifier1 = Identifier(x)
let identifier2: Identifier = Identifier("smth")
```
It seemed to me a bit odd overloading the initializer worked like that in this case. Is there a reason where a string literal could be used with the other `init?(_ token: TokenSyntax)`?

This behavior doesn't happen with an explicit argument label so that's why I changed that. Do you think there could be a better way to fix this?

**Comment by MAJKFL** [comment]
> How did you find that, Jakub?

We’ve discovered this with @DougGregor when validating `SwiftLexicalLookup` against `ASTScope`. We were trying out different edge cases and noticed this behavior. 

`SwiftLexicalLookup` handles dollar identifiers slightly differently for anonymous lookup than the compiler. When encountering a closure without parameters, the lookup just prompts clients with `LookupResult.mightIntroduceDollarIdentifiers` result flag.

For a lookup with a particular identifier, it does introduce a specific dollar identifier name (`LookupName.dollarIdentifier`), and that’s where we need its canonical representation to not introduce a monstrosity such as `$000000000042`.

**Comment by MAJKFL** [comment]
@swift-ci Please test Windows

**Review by DougGregor — APPROVED** [review]
Looks reasonable to me, thank you

File: Sources/SwiftSyntax/Identifier.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/Identifier.swift:56]: Oh, I see. `TokenSyntax` is expressible by string literal as well, that’s why it’s ambiguous.

I’m not a fan of the `staticString` label. How about we use the argument label to convey that the string needs to be canonical? That way a user will get the overload via `TokenSyntax` that canonicalizes the name by default and they need to opt into this faster but more restrictive overload.

Ie. change it to `init(canonicalName:)` or something of the sort.

Comment (MAJKFL) [reply Sources/SwiftSyntax/Identifier.swift]: Oh that makes sense why it didn't work with string literals, thanks for explanation. I also like the idea of `init(canonicalName:)`. I changed it.

**Comment by MAJKFL** [comment]
@swift-ci Please test

**Comment by MAJKFL** [comment]
@swift-ci Please test Windows

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by DougGregor** [comment]
@swift-ci Please test Windows
